### PR TITLE
Add KubeRay Manifests

### PR DESCRIPTION
**Pull request by MichaelClifford**
This PR consists of two main changes:

1) Addition of the manifests needed to install the [KubeRay Operator](https://github.com/ray-project/kuberay) along with a few tests.   
2) Removed the classic ray operator manifests. 

This update is required as the classic ray operator is being depreciated and losing support. Moving forward, KubeRay will be the standard way to manage ray clusters on Kubernetes systems.

### CODE REVIEW

**Comment by LaVLaS** [comment]
@MichaelClifford Is there any reason to keep the older ray operator?

**Comment by MichaelClifford** [comment]
@LaVLaS Ultimately, no.  Tests still need to be written for the KubeRay Operator. So, I was trying to break this down into a few smaller PR's, maintaining the old operator until the new one is fully ready to replace it. But,  if you'd prefer the entire update comes in 1 big PR. That's fine to. Let me know :)

File: ray/README.md
Comment (accorvin) [inline ray/README.md]: Can we add a note about how to determine `<Ray_Cluster_Service_Name>`

File: ray/operator/base/kustomization.yaml
Comment (accorvin) [inline ray/operator/base/kustomization.yaml]: I don't think we should block merging the PR on this, but I think we should think through a way to un-hardcode the namespace for this.

File: ray/README.md
Comment (accorvin) [inline ray/README.md]: This file appears to not exist.

File: ray/operator/base/kustomization.yaml
Comment (MichaelClifford) [reply ray/operator/base/kustomization.yaml]: @accorvin :+1:

The operator lives in its own namespace, so I don't think it should be installed inside of say, the ODH namespace. `ray-system` is the default name used by the KubeRay Project.  Is there an issue with it spawning another namespace?

File: ray/README.md
Comment (accorvin) [inline ray/README.md]: It occurs to me that these instructions bypass the ODH entirely. We're just manually installing kuberay from the raw kustomize artifacts. If we're providing a kfdef in this repo then IMO and stating that an ODH installation is a prerequisite then IMO our preferred instructions should be to apply that kfdef for installation.

File: ray/operator/base/kustomization.yaml
Comment (LaVLaS) [reply ray/operator/base/kustomization.yaml]: Using the `namespace:` in `kustomization.yaml` is the preferred method since it propagates to every namespace reference in the manifest and there are plugins we could use to add hooks that modify that functionality

Comment (accorvin) [reply ray/operator/base/kustomization.yaml]: Yeah so I think the only issue here is that we also create the namespace and the name is hardcoded there, so even if someone changed the namespace passed to kustomize it would still create the `ray-system` namespace.

File: ray/README.md
Comment (MichaelClifford) [reply ray/README.md]: That is a good point @accorvin.  I can update the README so that is assumes you've installed KubeRay via the KfDef, then just how to confirms its installed and working. Would that be better?

Comment (accorvin) [reply ray/README.md]: @MichaelClifford yeah, I like that plan.

**Comment by MichaelClifford** [comment]
@accorvin  Thanks for the review! Made some updates. Let me know if they address your notes correctly.

File: ray/README.md
Comment (accorvin) [inline ray/README.md]: @MichaelClifford I think it would be better to provide a valid kfdef in this repo under the `ray` directory. Then the instructions here become to oc create that kfdef file.

Comment (MichaelClifford) [reply ray/README.md]: I'm happy to to it that way, but the rest of the components in this repo provide only the KustomizeConfig snippet. And the valid KfDef with all the components is under the 'KfDef' directory. 

So that changes would introduce a different format for the Ray directory. 

@LaVLaS Do you mind if we make this sort of change? I don't care either way, but wanted to point it out.

**Comment by MichaelClifford** [comment]
@accorvin @LaVLaS  
Bump :smile: 

Is this good to go, or should we add a minimal KfDef to here?

**Comment by accorvin** [comment]
@MichaelClifford to me this repo would be much more valuable if we did include a minimal KFDef that is Ray-specific. 

I'm expecting a lot of people we point at this repo will have no idea what the heck a kfdef is, so the current state of the readme isn't super useful. If we point the user to the full odh-contrib kfdef, this will deploy a bunch of stuff that the customer doesn't need or care about. Inevitably this will include things that fail to deploy correctly. This will probably just annoy people.

Check out the installation instructions that the data science pipelines team just wrote: https://github.com/opendatahub-io/data-science-pipelines/tree/master/manifests/opendatahub#installation-steps I think that does a pretty good job of creating a very easy to follow set up steps for someone who has never used the ODH before.

File: ray/README.md
Comment (LaVLaS) [reply ray/README.md]: IMO, the README should have two instructions for deployment
1. Standalone deployment by running `kustomize` on the manifest in the repo
2. Instructions for deploying via the kfdef.  

If there are required ODH dependencies that Ray needs then I would say a kfdef file is required.

File: ray/operator/base/kustomization.yaml
Comment (LaVLaS) [inline ray/operator/base/kustomization.yaml]: @MichaelClifford Is there any concern with referencing a `nightly` tag breaking any deployments?

Comment (MichaelClifford) [reply ray/operator/base/kustomization.yaml]: Good point @LaVLaS This is what the KubeRay repo suggests, but we can pin this down to a recent version.

File: ray/operator/base/kustomization.yaml
Comment (LaVLaS) [reply ray/operator/base/kustomization.yaml]: So that is an issue with the use [kustomize v3.2.0](https://github.com/opendatahub-io/opendatahub-operator/blob/master/go.mod#L61) in the current operator code where it does not run the `namespace:` transformer on `kind: Namespace` objects.

That issue will be resolved when we update the kustomize version in the operator to kustomize >= [v3.8.3](https://github.com/kubernetes-sigs/kustomize/pull/2953) this month.

@accorvin I'll defer to you if you want remove the Namespace creation.

Comment (accorvin) [reply ray/operator/base/kustomization.yaml]: I think if we keep the namespace creation here then we should include basic instructions for how to change/override the namespace. IMO it's better to make namespace creation a prerequisite step that the user must do.

File: ray/README.md
Comment (accorvin) [inline ray/README.md]: You have a typo here "requires that  to have". I think this should be something like "requires that you have"

File: ray/README.md
Comment (accorvin) [reply ray/README.md]: I like this now.

**Comment by accorvin** [comment]
I'll work on testing this this morning.

**Comment by MichaelClifford** [comment]
thanks @accorvin I'll fix that typo now :)

File: ray/README.md
Comment (accorvin) [inline ray/README.md]: These instructions don't work when using the minimal kfdef. There is no ray-system namespace created. If this is supposed to be the opendatahub namespace, there is a pod called `ray-operator-...`, but not kuberay.

File: ray/README.md
Comment (accorvin) [inline ray/README.md:83]: @MichaelClifford I don't think this should block merging of this PR, but I got this error in my notebook:
```
[RuntimeError: Client Ray installation incompatible with server: client is 2022-12-06, server is 2022-10-05](ray://kuberay-cluster-test-head-svc:10001'</span><span)
```

This was after doing a simple `!pip install ray` in the notebook using the out of the box standard data science ODH notebook image.

We may want to upgrade the default Ray cluster node images we use to be more up to date.

**Review by accorvin — APPROVED** [review]

File: ray/README.md
Comment (MichaelClifford) [reply ray/README.md:83]: Thanks @accorvin, 

Yeah the version of Ray in the notebook env has to match that in the worker nodes. I think I made the assumption here that the user would be using the "codeflare notebook" that has the matching ray version.

Comment (accorvin) [reply ray/README.md:83]: Makes sense. We should document that assumption.

Comment (MichaelClifford) [reply ray/README.md:83]: For now I can add a note here that says which version of Ray to install

Comment (accorvin) [reply ray/README.md:83]: 🚢

**Comment by accorvin** [comment]
Changes look good to me.

**Pull request merged**
